File: frontend/src/types/factor.ts
export interface FactorDefinition {
  id: string;
  user_id: string | null;
  name: string;
  display_name: string;
  description?: string;
  expression: string;
  category: 'system' | 'custom' | 'technical' | 'fundamental';
  parameters: Record<string, unknown>;
  is_active: boolean;
  created_at: string;
}

export interface FactorAnalysis {
  factor_name: string;
  mean_ic: number;
  ic_ir: number;
  ic_std: number;
  positive_ic_ratio: number;
  num_stocks: number;
  ic_series: number[];
}

File: backend/app/schemas/factor.py
from pydantic import BaseModel, ConfigDict, field_validator
from typing import Optional
from uuid import UUID
from datetime import datetime


class FactorCreate(BaseModel):
    name: str
    display_name: str
    description: Optional[str] = None
    expression: str
    category: str = "custom"
    parameters: dict = {}

    @field_validator("name")
    @classmethod
    def name_valid(cls, v: str) -> str:
        if not v.replace("_", "").isalnum():
            raise ValueError("因子名称只能包含字母、数字和下划线")
        return v


class FactorUpdate(BaseModel):
    display_name: Optional[str] = None
    description: Optional[str] = None
    expression: Optional[str] = None
    parameters: Optional[dict] = None
    is_active: Optional[bool] = None


class FactorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: Optional[UUID] = None
    name: str
    display_name: str
    description: Optional[str] = None
    expression: str
    category: str
    parameters: dict
    is_active: bool
    created_at: datetime


class FactorPreviewRequest(BaseModel):
    expression: str
    stock_code: str
    start_date: str
    end_date: str


class FactorAnalysisRequest(BaseModel):
    factor_id: UUID
    stock_codes: list[str]
    start_date: str
    end_date: str
